### PR TITLE
Title: Unselect title by blur event (B)

### PR DESCRIPTION
**Pull request by ephox-mogran**
This is building on https://github.com/WordPress/gutenberg/pull/2948

## Description

Add some blur and focus handlers to the outer div of the title post so that we can detect focus has transferred outside of the container. Built on the previous PR by allowing the Copy Permalink button to be clicked.

## How Has This Been Tested?

I have manually tested it but did not know how to write unit tests for it.

## Types of changes

This change is a local change to the post-title component, as well as a small change to Permalink. It requires an additional prop for Permalinks, but they are only used in Titles.

## Checklist:
- [X] My code is tested (manually)
- [X] My code follows the WordPress code style.
- [X] My code follows has proper inline documentation (there was no documentation in the file)

### CODE REVIEW

**Comment by ephox-mogran** [comment]
Summary of changes

* added blur and focus handlers to the outer div, so that focusing something inside (like the button) does not fire a blur event
* focused the textarea after clicking copy. On of the issue with the clipboard button is that it uses a library which creates a textarea and focuses it to copy to the clipboard. That textarea is then removed while it is the active element, which makes the browser fallback to making body the active element. This fired a blur and made the unselect fire. By focusing the title textarea after copying, we avoid this.
* added the state: hasFocusWithin which must also be true (as well as isSelected which seems to revolve around also detecting typing) for the title 'block' decorations to appear.

File: editor/post-title/index.js [hunk 1]
```diff
@@ -76,6 +89,26 @@ class PostTitle extends Component {
...
+
+	onOuterBlur( e ) {
+		const target = e.currentTarget;
+		clearTimeout( this.blurTimer );
+		this.blurTimer = setTimeout( () => {
+			this.setFocused( target.contains( document.activeElement ) );
+		}, 0 );
+	}
+
 	handleClickOutside() {
```
Comment (ephox-mogran) [inline editor/post-title/index.js]: We can probably delete this now, right?

File: editor/post-title/index.js [hunk 2]
```diff
@@ -88,12 +121,15 @@ class PostTitle extends Component {
 
 	render() {
 		const { title } = this.props;
-		const { isSelected } = this.state;
-		const className = classnames( 'editor-post-title', { 'is-selected': isSelected } );
```
Comment (ephox-mogran) [inline editor/post-title/index.js]: Maybe isSelected should be something like isTyping?

Comment (aduth) [reply editor/post-title/index.js]: The naming had confused me as well when I first encountered it. I think `isTyping` would be an improvement, yes.

File: editor/post-title/index.js [hunk 3]
```diff
@@ -88,12 +121,15 @@ class PostTitle extends Component {
...
+		const className = classnames( 'editor-post-title', { 'is-selected': isSelected && hasFocusWithin } );
 
 		return (
-			<div className={ className }>
-				{ isSelected && <PostPermalink /> }
+			<div
+				className={ className }
+				onBlur={ this.onOuterBlur }
+				onFocus={ this.onOuterFocus }>
+				{ isSelected && hasFocusWithin && <PostPermalink onLinkCopied={ this.focusText } /> }
```
Comment (aduth) [inline editor/post-title/index.js]: You mentioned the clipboard button library being the cause that led to this implementation to focus text again after copying. Is that something that can be (or perhaps is already) fixed at the library? Or fixed on the clipboard button? Maybe using our "withFocusReturn" higher-order component could work as a solution?

https://github.com/WordPress/gutenberg/blob/master/components/higher-order/with-focus-return/index.js

File: editor/post-title/index.js [hunk 4]
```diff
@@ -76,6 +89,26 @@ class PostTitle extends Component {
...
+
+	onOuterFocus() {
+		clearTimeout( this.blurTimer );
+		this.setFocused( true );
+	}
+
+	onOuterBlur( e ) {
+		const target = e.currentTarget;
+		clearTimeout( this.blurTimer );
+		this.blurTimer = setTimeout( () => {
```
Comment (aduth) [inline editor/post-title/index.js]: `setTimeout` is a code smell in components and can often lead to difficult-to-maintain components. It's often a signal that we don't have a full understanding of the natural flow of events and deferring is a convenient waiting mechanism. I'm wondering if there is a better way to implement this.

File: editor/post-title/index.js [hunk 5]
```diff
@@ -35,17 +34,27 @@ class PostTitle extends Component {
...
 		this.state = {
 			isSelected: false,
+			hasFocusWithin: false,
 		};
+		this.blurTimer = null;
 	}
 
 	componentDidMount() {
 		document.addEventListener( 'selectionchange', this.onSelectionChange );
+		// eslint-disable-next-line react/no-find-dom-node
```
Comment (aduth) [inline editor/post-title/index.js]: The lint rule is correct here: `findDOMNode` should be discouraged, and this specific behavior can be recreated with a [`ref`](https://reactjs.org/docs/refs-and-the-dom.html). More ideally, we'd not rely on the DOM to determine how to assign focus state (the other way around being the preference, state -> DOM).

Also worth noting that this will always cause a re-render immediately after the first render, even if state is not changing (`setState` will always cause render, even if the `hasFocusWithin` is and stays `false`).

When would we expect this element to be the active element after a mount?

File: editor/post-title/index.js (hunk 1 above)
Comment (aduth) [reply editor/post-title/index.js]: Yes, `clickOutside` relies on this, so if we're removing it, it's likely become unused.

File: editor/post-title/index.js (hunk 5 above)
Comment (ephox-mogran) [reply editor/post-title/index.js]: Yeah, I'm new to React so I wasn't sure if this would ever be true. I guess it would be true if the input that appeared inside it had autofocus, but maybe not exactly on mounting. I'm happy to set it to false on startup and have nothing here.

File: editor/post-title/index.js (hunk 4 above)
Comment (ephox-mogran) [reply editor/post-title/index.js]: Yeah, I have the setTimeout too, however I can't see a current way around it. Maybe you can. I'll explain what I think are the sequence of events

a) you focus inside the post title (fires focus event)
b) you click the copied button
  - fires a blur first on the post title (which would cause unselect)
  - then fires a focus on the copy button (based on the mousedown)
  - then fires a click on the copy button (which is what clipboard library is relying on)
  - this then creates a fake textarea and focuses it, does the copying magic and then removes it from the DOM
  - then I return focus to the textarea, because now the focused element is removed from the DOM. I think you are right, though, that we could use the returnFocus HOC.

The big problem is that blur fires before the next DOM element is focused. This is an issue with most frameworks. Some browsers might supply a relatedTarget on the blur event, but not all of them ... which has typically led to setTimeout solutions abounding like https://medium.com/@jessebeach/dealing-with-focus-and-blur-in-a-composite-widget-in-react-90d3c3b49a9b and https://gist.github.com/pstoica/4323d3e6e37e8a23dd59 and https://stackoverflow.com/questions/11592966/get-the-newly-focussed-element-if-any-from-the-onblur-event/11592974#11592974. 

However, there might be a way. In our previous dealings with these issues, we've often had to resort to a setTimeout. Hopefully, you're aware of a better option.

File: editor/post-title/index.js (hunk 3 above)
Comment (ephox-mogran) [reply editor/post-title/index.js]: I think that might work, but now that I think about it, doesn't this only work on componentDidUnmount? The only react controlled component is the copy button (the textarea that gets focus is created by the clipboard library itself), and I thought the copy button wasn't being removed from the DOM after clicking on it?

Comment (ephox-mogran) [reply editor/post-title/index.js]: There is also an option in the clipboard library of passing through trigger, and then calling clearSelection (which will focus trigger and then remove the selection), but it would be just calling focus outside of the react in the exact same way. Is that actually preferable?

**Comment by ephox-mogran** [comment]
I've had another go at this which is working reasonably well. Essentially it uses a similar approach to clickOutside but for focus. I ignore blur events, and just use focusOutside to simulate the same thing. Hoping to update the PR soon.

**Comment by ephox-mogran** [comment]
@aduth what about this? I don't have particularly extensive tests, but I'm not really sure how to write them. Ideally, I'd want to create a hierarchy of components and some outside the component, and shift focus around and assert when the handleFocusOutside method fires.

File: components/clipboard-button/index.js [hunk 6]
```diff
@@ -15,12 +15,16 @@ import { findDOMNode, Component } from '@wordpress/element';
  */
 import { Button } from '../';
 
-class ClipboardButton extends Component {
+// This creates a container to put the textarea in which isn't removed by react
+// If react removes the textarea first, then the clipboard fails when trying to remove it
+class ClipboardContainer extends Component {
```
Comment (aduth) [inline components/clipboard-button/index.js]: Does this need to be a separate component, or can we assign the button of `ClipboardButton` as the `container`?

File: components/higher-order/with-focus-outside/index.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+			) {
+				this.__wrappedInstance.handleFocusOutside( e );
+			}
+		}
+
+		bindRef( wrappedRefCallback, ref ) {
+			this.__wrappedInstance = ref;
+			// eslint-disable-next-line react/no-find-dom-node
+			this.__domNode = findDOMNode( ref );
+			if ( wrappedRefCallback ) {
```
Comment (aduth) [inline components/higher-order/with-focus-outside/index.js]: What is the purpose of the `wrappedRefCallback`? Is this just inherited from click outside? Probably best to avoid introducing features unless we foresee needing them.

File: editor/post-title/index.js [hunk 8]
```diff
@@ -76,8 +85,19 @@ class PostTitle extends Component {
 		this.setState( { isSelected: false } );
 	}
 
-	handleClickOutside() {
-		this.setState( { isSelected: false } );
+	setFocused( focused ) {
+		this.setState( { hasFocusWithin: focused } );
+		if ( focused ) {
+			this.props.clearSelectedBlock();
```
Comment (aduth) [inline editor/post-title/index.js]: Why is this needed? Ideally the title focus handler shouldn't need to deal with block selection (i.e. it should be block focus leave events clearing its selected state).

File: editor/post-title/index.js [hunk 9]
```diff
@@ -76,8 +85,19 @@ class PostTitle extends Component {
...
+		}
+	}
+
+	handleFocusOutside( ) {
```
Comment (aduth) [inline editor/post-title/index.js]: Styling note: There should be no spaces within the parentheses:

>No filler spaces in empty constructs (e.g., `{}`, `[]`, `fn()`).

https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#spacing

I'll plan to look and see if we can enforce this by ESLint.

**Review by aduth — COMMENTED** [review]
Hmm, I'm wondering if the `withFocusOutside` higher-order component is necessary, or if it might be sufficient to fix the clipboard button container and move the `onBlur` introduced in #2948 from the `Textarea` to the root wrapping `div` (accounting for focus within both the textarea and post permalink).

File: components/clipboard-button/index.js (hunk 6 above)
Comment (ephox-mogran) [reply components/clipboard-button/index.js]: It needs to have shouldUpdateFalse. The problem is if it gets redrawn, then the textarea that is inside gets removed, and the clipboard library throws an error when trying to remove it. We just need a component that can add things added to it, and isn't going to get recreated.

File: components/higher-order/with-focus-outside/index.js (hunk 7 above)
Comment (ephox-mogran) [reply components/higher-order/with-focus-outside/index.js]: Yeah, based on the click outside. I can remove it.

File: editor/post-title/index.js (hunk 8 above)
Comment (ephox-mogran) [reply editor/post-title/index.js]: When I was testing it before, that wasn't working. But that was an earlier version. I'm happy to remove it and try the blur of the other blocks.

File: editor/post-title/index.js (hunk 9 above)
Comment (ephox-mogran) [reply editor/post-title/index.js]: Noted.

File: components/clipboard-button/index.js (hunk 6 above)
Comment (ephox-mogran) [reply components/clipboard-button/index.js]: I had it as the button originally, and kept getting an error in the clipboard library when it tried to remove the textarea, because react had already removed it because the button had been redrawn. You would get this error if you tried clicking on the button after you had already clicked on it. Therefore, the idea was to make a component that could have anything inside it, and react wouldn't keep trying to remove it. It's a limitation of using a third party library which appends components into the DOM. Does that make sense?

File: editor/post-title/index.js (hunk 8 above)
Comment (ephox-mogran) [reply editor/post-title/index.js]: Removed this.

File: editor/post-title/index.js (hunk 9 above)
Comment (ephox-mogran) [reply editor/post-title/index.js]: Fixed.

File: components/higher-order/with-focus-outside/index.js (hunk 7 above)
Comment (ephox-mogran) [reply components/higher-order/with-focus-outside/index.js]: Removed this extra wrapping.

File: components/clipboard-button/index.js (hunk 6 above)
Comment (ephox-mogran) [reply components/clipboard-button/index.js]: If you don't supply container, then it defaults to body, which is fine normally, but it will fire a blur which will trigger focus outside. That's why I'm supplying container ... to keep the focus inside the post title block.
